runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java

```diff
@@ -18,107 +18,230 @@
 
-  private StreamingRuntimeContext runtime = null;
-  private StreamSource.ManualWatermarkContext<WindowedValue<T>> context = null;
+  /**
+   * Keep the options so that we can initialize the readers.
+   */
+  private final SerializedPipelineOptions serializedOptions;
 
-  private volatile boolean isRunning = false;
+  /**
+   * For snapshot and restore.
+   */
+  private final Coder<CheckpointMarkT> checkpointMarkCoder;
 
-  private final SerializedPipelineOptions serializedOptions;
+  /**
+   * The split sources. We split them in the constructor to ensure that all parallel
+   * sources are consistent about the split sources.
+   */
+  private List<? extends UnboundedSource<OutputT, CheckpointMarkT>> splitSources;
 
-  /** Instantiated during runtime **/
-  private transient UnboundedSource.UnboundedReader<T> reader;
+  /**
+   * Make it a field so that we can access it in {@link #trigger(long)} for
+   * emitting watermarks.
+   */
+  private transient List<UnboundedSource.UnboundedReader<OutputT>> readers;
 
-  public UnboundedSourceWrapper(PipelineOptions pipelineOptions, Read.Unbounded<T> transform) {
-    this.name = transform.getName();
-    this.serializedOptions = new SerializedPipelineOptions(pipelineOptions);
-    this.source = transform.getSource();
-  }
+  /**
+   * Initialize here and not in run() to prevent races where we cancel a job before run() is
+   * ever called or run() is called after cancel().
+   */
+  private volatile boolean isRunning = true;
 
-  public String getName() {
-    return this.name;
-  }
+  /**
+   * Make it a field so that we can access it in {@link #trigger(long)} for registering new
+   * triggers.
+   */
+  private transient StreamingRuntimeContext runtimeContext = null;
+
+  /**
+   * Make it a field so that we can access it in {@link #trigger(long)} for emitting
+   * watermarks.
+   */
+  private transient StreamSource.ManualWatermarkContext<WindowedValue<OutputT>> context = null;
+
+  private transient byte[] restoredState = null;
+
+  public UnboundedSourceWrapper(
+      PipelineOptions pipelineOptions,
+      UnboundedSource<OutputT, CheckpointMarkT> source,
+      int parallelism) throws Exception {
+    this.serializedOptions = new SerializedPipelineOptions(pipelineOptions);
 
-  WindowedValue<T> makeWindowedValue(T output, Instant timestamp) {
-    if (timestamp == null) {
-      timestamp = BoundedWindow.TIMESTAMP_MIN_VALUE;
+    if (source.requiresDeduping()) {
+      LOG.warn("Source {} requires deduping but Flink runner doesn't support this yet.", source);
     }
-    return WindowedValue.of(output, timestamp, GlobalWindow.INSTANCE, PaneInfo.NO_FIRING);
+
+    checkpointMarkCoder = source.getCheckpointMarkCoder();
+    // get the splits early. we assume that the generated splits are stable,
+    // this is necessary so that the mapping of state to source is correct
+    // when restoring
+    splitSources = source.generateInitialSplits(parallelism, pipelineOptions);
```

Comment:
I'm trying to understand what this is doing. Here's the workflow from my POV:

* at submission time, user has specified some UnboundedSource in the graph. For example, `CountingInput.unbounded()` [which generates the non-negative integers `[0, 1, 2, ...)`].
* at submission time, we decide an upper bound on the number of splits. So, say, 11. We will then split the source that many ways, into `[0, 11, 22, ...)`, `[1, 12, 23, ...)`, ..., `[10, 21, 32, ...)`.

* at job runtime, we spin up workers and make some assignment of splits to workers. Supposing we have 5 workers and use round-robin, each worker will get 2-3 splits of the 11.
* the runner harness then asks each worker to start its sources and produce readers from `null` checkpoints.
* Potentially asynchronously and independently across workers, the system will ask a reader to checkpoint in order to commit some work. The local worker is required to persist that checkpoint into its local state.
* Suppose a worker fails. Then we bring it back up, but instead of giving it a `null` checkpoint we give it the last checkpoint we persisted. Thus we're fault tolerant.

My main point is that I don't understand the questions here. if `N` initial splits, should have `N` sources, `N` readers, and `N` checkpoints. If the user changes the number of VMs in a job, they probably are increasing it and those new VMs will just not have any splits. (Unless the system rebalances, aka sends a `<Split Source, Checkpoint>` pair to a new machine).

Does that make sense? Also happy to jump on a call for realtime convo.